src/libraries/componentdirectorywatcher.h

```diff
@@ -0,0 +1,72 @@
+/****************************************************************************
+**
+** Copyright (C) 2017 N7 Mobile sp. z o. o.
+** Contact: http://n7mobile.pl/Space
+**
+** This file is part of ASN.1/ACN Plugin for QtCreator.
+**
+** Plugin was developed under a programme and funded by
+** European Space Agency.
+**
+** This Plugin is free software: you can redistribute it and/or modify
+** it under the terms of the GNU General Public License as published by
+** the Free Software Foundation, either version 3 of the License, or
+** (at your option) any later version.
+**
+** This Plugin is distributed in the hope that it will be useful,
+** but WITHOUT ANY WARRANTY; without even the implied warranty of
+** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+** GNU General Public License for more details.
+**
+** You should have received a copy of the GNU General Public License
+** along with this program.  If not, see <http://www.gnu.org/licenses/>.
+**
+****************************************************************************/
+
+#pragma once
+
+#include <memory>
+
+#include <QObject>
+#include <QTimer>
+#include <QFileSystemWatcher>
+
+#include <settings/libraries.h>
+
+#include "componentlibrarydispatcher.h"
+
+namespace Asn1Acn {
+namespace Internal {
+namespace Libraries {
+
+class ComponentDirectoryWatcher : public QObject
+{
+    Q_OBJECT
+
+public:
+    ComponentDirectoryWatcher(Settings::LibrariesConstPtr libraries,
+                              std::unique_ptr<CompontentLibraryDispatcher> dispatcher,
+                              QObject *parent = nullptr);
+
+private slots:
+    void configChanged();
+    void filesChanged(const QString &path);
+
+    void resetWatch();
+
+private:
+    void addAllLibraries();
+    void addMainDirectory(const QString &path);
+    void addSubDirectory(const QString &path);
+    void removeAll();
+
+    std::unique_ptr<QFileSystemWatcher> m_fsWatcher;
+    std::unique_ptr<CompontentLibraryDispatcher> m_dispatcher;
```

Comment:
The class watch as follows: it receives list of all currently watched directories and list of all files from file watcher. Than what it does is to match files with directories, and trigger parsing for each files collection created this way. I came up with the name dispatching for the process. Do you think it should be changed?

Reply:
I just usually name `dispatcher` things that also perform some form of "routing" - dispatching to several different modules. Name is just used a lot in other project and it triggered me a little ;)

Can it go directly to Processor?

Reply:
In current form it can not, as all the files are stored in one collection, so some sorting needs to be performed, to find out which metadata files belong to the same library. To pass it directly to processor, the sorting functionality would need to be moved to _ComponentDirectoryWatcher_ class in some way, for instance thorough adding the files paths to something like `std::map<QString, QStringList>`.

Reply:
ok, keep it as is